### PR TITLE
RMNHome

### PR DESCRIPTION
**Pull request by RensR**
New PR to have this be standalone

### CODE REVIEW

File: contracts/src/v0.8/ccip/rmn/RMNHome.sol [hunk 1]
```diff
@@ -8,145 +8,314 @@ import {OwnerIsCreator} from "../../shared/access/OwnerIsCreator.sol";
...
+  VersionedConfig[MAX_CONCURRENT_CONFIGS] private s_configs;
 
-        // minObservers are tenable
-        if (!(newConfig.sourceChains[i].minObservers <= newConfig.sourceChains[i].observerNodeIndices.length)) {
-          revert MinObserversTooHigh();
-        }
-      }
-    }
+  /// @notice The total number of configs ever set, used for generating the version of the configs.
+  uint32 private s_configCount = 0;
```
Comment (connorwstein) [inline contracts/src/v0.8/ccip/rmn/RMNHome.sol]: s_currentVersion

File: contracts/src/v0.8/ccip/rmn/RMNHome.sol [hunk 2]
```diff
@@ -8,145 +8,314 @@ import {OwnerIsCreator} from "../../shared/access/OwnerIsCreator.sol";
...
+  /// - The activeConfigIndex is flipped.
+  /// - The digest of the old active config is deleted.
+  function promoteCandidateAndRevokeActive(bytes32 digestToPromote, bytes32 digestToRevoke) external onlyOwner {
+    uint256 candidateConfigIndex = s_activeConfigIndex ^ 1;
+    if (s_configs[candidateConfigIndex].configDigest != digestToPromote) {
+      revert ConfigDigestMismatch(s_configs[candidateConfigIndex].configDigest, digestToPromote);
+    }
+
+    uint256 activeConfigIndex = s_activeConfigIndex;
+    if (s_configs[activeConfigIndex].configDigest != digestToRevoke) {
```
Comment (0xsuryansh) [inline contracts/src/v0.8/ccip/rmn/RMNHome.sol]: nit

```suggestion
    VersionedConfig storage activeConfig = s_configs[activeConfigIndex];
    if (activeConfig.configDigest != digestToRevoke) {
```

Comment (0xsuryansh) [reply contracts/src/v0.8/ccip/rmn/RMNHome.sol]: Cleaner and saves little gas, 

event to : `ConfigDigestMismatch(activeConfig.configDigest, digestToRevoke);`

similarly `delete s_configs[activeConfigIndex].configDigest;` can be changed to `delete activeConfig.configDigest;`

File: contracts/src/v0.8/ccip/rmn/RMNHome.sol [hunk 3]
```diff
@@ -8,145 +8,314 @@ import {OwnerIsCreator} from "../../shared/access/OwnerIsCreator.sol";
...
+  /// @notice Returns the current active and candidate config digests.
+  /// @dev Can be bytes32(0) if no config has been set yet or it has been revoked.
+  /// @return activeConfigDigest The digest of the active config.
+  /// @return candidateConfigDigest The digest of the candidate config.
+  function getConfigDigests() external view returns (bytes32 activeConfigDigest, bytes32 candidateConfigDigest) {
+    return (s_configs[s_activeConfigIndex].configDigest, s_configs[s_activeConfigIndex ^ 1].configDigest);
+  }
+
+  /// @notice Returns the active config digest
+  function getActiveDigest() public view returns (bytes32) {
```
Comment (0xsuryansh) [inline contracts/src/v0.8/ccip/rmn/RMNHome.sol]:
```suggestion
  function getActiveDigest() external view returns (bytes32) {
```

**Review by connorwstein — COMMENTED** [review]
Think we should allow only state transitions we want:
- init->candidate
- candidate->active
- active->init
- active->active/candidate
- active/candidate->active 
- active/candidate->active/candidate

then revert on everything else

File: contracts/src/v0.8/ccip/rmn/RMNHome.sol [hunk 4]
```diff
@@ -7,146 +7,356 @@ import {OwnerIsCreator} from "../../shared/access/OwnerIsCreator.sol";
...
+    bytes32 offchainPublicKey; // Observations are signed with this public key, and are only verified offchain.
   }
 
   struct SourceChain {
-    uint64 chainSelector;
-    uint64[] observerNodeIndices; // indices into Config.nodes, strictly increasing
-    uint64 minObservers; // required to agree on an observation for this source chain
+    uint64 chainSelector; // ─────╮ The Source chain selector.
+    uint64 minObservers; // ──────╯ Required number of observers to agree on an observation for this source chain.
+    uint256 observerNodesBitmap; // ObserverNodesBitmap & (1<<i) == (1<<i) iff nodes[i] is an observer for this source chain.
```
Comment (makramkd) [inline contracts/src/v0.8/ccip/rmn/RMNHome.sol]:
```suggestion
    uint256 observerNodesBitmap; // ObserverNodesBitmap & (1<<i) == (1<<i) iff StaticConfig.nodes[i] is an observer for this source chain.
```

File: contracts/src/v0.8/ccip/rmn/RMNHome.sol [hunk 5]
```diff
@@ -7,146 +7,356 @@ import {OwnerIsCreator} from "../../shared/access/OwnerIsCreator.sol";
...
+  /// @notice Returns the candidate config digest
+  function getCandidateDigest() public view returns (bytes32) {
+    return s_configs[s_activeConfigIndex ^ 1].configDigest;
+  }
+
+  /// @notice The offchain code can use this to fetch an old config which might still be in use by some remotes. Use
+  /// in case one of the configs is too large to be returnable by one of the other getters.
+  /// @param configDigest The digest of the config to fetch.
+  /// @return versionedConfig The config and its version.
+  /// @return ok True if the config was found, false otherwise.
```
Comment (makramkd) [inline contracts/src/v0.8/ccip/rmn/RMNHome.sol:163]: Just wondering if this is needed, won't `configDigest` on the `VersionedConfig` be `ZERO_DIGEST` which is always an invalid digest and hence config does not exist?

File: contracts/src/v0.8/ccip/rmn/RMNHome.sol [hunk 6]
```diff
@@ -7,146 +7,356 @@ import {OwnerIsCreator} from "../../shared/access/OwnerIsCreator.sol";
...
+    return s_configs[s_activeConfigIndex].configDigest;
+  }
 
-    for (uint256 i = 0; i < CONFIG_RING_BUFFER_SIZE; ++i) {
-      if ((i == newConfigIndex || revokePastConfigs) && s_configCounts[i] > 0) {
-        emit ConfigRevoked(_configDigest(VersionedConfig({version: s_configCounts[i], config: s_configs[i]})));
-        delete s_configCounts[i];
+  /// @notice Returns the candidate config digest
+  function getCandidateDigest() public view returns (bytes32) {
+    return s_configs[s_activeConfigIndex ^ 1].configDigest;
```
Comment (makramkd) [inline contracts/src/v0.8/ccip/rmn/RMNHome.sol]: I dunno if this'd be inefficient in solidity (or maybe the call would be inlined - hopefully) but extracting the `s_activeConfigIndex ^ (0|1)` into named funcs might be helpful for readability, e.g

```solidity
function activeIndex() private pure returns (uint256) { return s_activeConfigIndex ^ 0; }
function candidateIndex() private pure returns (uint256) { return s_activeConfigIndex ^ 1; }

// usage
s_configs[activeIndex()].configDigest; // etc.
```

File: contracts/src/v0.8/ccip/rmn/RMNHome.sol [hunk 7]
```diff
@@ -7,146 +7,356 @@ import {OwnerIsCreator} from "../../shared/access/OwnerIsCreator.sol";
...
+/// This means the following calls are not allowed at the following states:
+/// - Init: promoteCandidateAndRevokeActive(), as there is no config to promote.
+/// - Init: revokeCandidate(), as there is no config to revoke
+/// - Active: revokeCandidate(), as there is no candidate to revoke
+/// Note that we explicitly do allow promoteCandidateAndRevokeActive() to be called when there is an active config but
+/// no candidate config. This is the only way to remove the active config. The alternative would be to set some unusable
+/// config as candidate and promote that, but fully clearing it is cleaner.
 contract RMNHome is OwnerIsCreator, ITypeAndVersion {
+  event ConfigSet(bytes32 indexed configDigest, uint32 version, StaticConfig staticConfig, DynamicConfig dynamicConfig);
+  event ConfigRevoked(bytes32 indexed configDigest);
```
Comment (makramkd) [inline contracts/src/v0.8/ccip/rmn/RMNHome.sol]: Do we care about being more specific about which config is revoked? We can revoke both active and candidate configs.

File: contracts/src/v0.8/ccip/rmn/RMNHome.sol [hunk 8]
```diff
@@ -7,146 +7,356 @@ import {OwnerIsCreator} from "../../shared/access/OwnerIsCreator.sol";
...
+  /// @notice Calculates the config digest for a given plugin key, static config, and version.
+  /// @param staticConfig The static part of the config.
+  /// @param version The version of the config.
+  /// @return The calculated config digest.
+  function _calculateConfigDigest(bytes memory staticConfig, uint32 version) internal view returns (bytes32) {
+    return bytes32(
+      (PREFIX & PREFIX_MASK)
+        | (
+          uint256(
+            keccak256(bytes.concat(abi.encode(bytes32("EVM"), block.chainid, address(this), version), staticConfig))
```
Comment (makramkd) [inline contracts/src/v0.8/ccip/rmn/RMNHome.sol:314]: Why is `staticConfig` `bytes.concat`'d and not `abi.encode`'d?

Comment (RensR) [reply contracts/src/v0.8/ccip/rmn/RMNHome.sol:314]: Both components are already abi.encoded. Afaik this should not compromise anything as the first section is all static length anyway

File: contracts/src/v0.8/ccip/rmn/RMNHome.sol (hunk 5 above)
Comment (RensR) [reply contracts/src/v0.8/ccip/rmn/RMNHome.sol:163]: That is true, but using the bool is more explicit. It's not strictly needed but I kinda like it.

**Review by makramkd — APPROVED** [review]

File: contracts/src/v0.8/ccip/rmn/RMNHome.sol (hunk 6 above)
Comment (connorwstein) [reply contracts/src/v0.8/ccip/rmn/RMNHome.sol]: agree, does seem easy to forget a xor somewhere. Can be a follow up though

File: contracts/src/v0.8/ccip/rmn/RMNHome.sol [hunk 9]
```diff
@@ -7,146 +7,358 @@ import {OwnerIsCreator} from "../../shared/access/OwnerIsCreator.sol";
...
+/// - Active -> Init: promoteCandidateAndRevokeActive()
+/// - ActiveAndCandidate -> Active: promoteCandidateAndRevokeActive()
+/// - ActiveAndCandidate -> Active: revokeCandidate()
+/// - ActiveAndCandidate -> ActiveAndCandidate: setCandidate()
+///
+/// This means the following calls are not allowed at the following states:
+/// - Init: promoteCandidateAndRevokeActive(), as there is no config to promote.
+/// - Init: revokeCandidate(), as there is no config to revoke
+/// - Active: revokeCandidate(), as there is no candidate to revoke
+/// Note that we explicitly do allow promoteCandidateAndRevokeActive() to be called when there is an active config but
```
Comment (connorwstein) [inline contracts/src/v0.8/ccip/rmn/RMNHome.sol:38]: I'd maybe add one more note (for completeness), that the other state transitions  ActiveAndCandidate->Init, Candidate -> Init are not possible in one step given the transition functions available. Alternatively would be more clear to have a little ascii diagram

**Review by connorwstein — APPROVED** [review]

File: contracts/src/v0.8/ccip/rmn/RMNHome.sol (hunk 9 above)
Comment (makramkd) [reply contracts/src/v0.8/ccip/rmn/RMNHome.sol:38]: +1 to cute ascii diagram with state transitions.

**Pull request merged**